### ISSUE TITLE
Corrupted Values When Using Deformations

### ISSUE THREAD

**Issue by mxalhussaini** [JuliaIBPM/ViscousFlow.jl#93]
Hello Professor, I appreciate the time and effort you invested in developing this useful and nice package.

While simulating cases with deformations, I came across a potential bug. The simulation involves morphing a NACA0006 into NACA0012 at zero angle of attack. I defined the morphing function  by extending the AbstractDeformationMotion as shown in the code below.

The values of Cd didn't look right when computed after finishing the time loop all at once. There is a difference in Cd values when computed at the corresponding time step versus post calculations. The values post calculations seem corrupted and changed compared when saved at the corresponding time during the time marching loop. I have checked the velocity and pressure fields at selected times and found that the velocity field is the same, but the pressure field has changed. Also note that this change of values occur at all times before the end of deformation time at 10 in this example.

The deformation used here is a morphing of an airfoil from NACA0006 to NACA0012 during the time from 5 to 10 as shown in the animation

![plot_32](https://github.com/JuliaIBPM/ViscousFlow.jl/assets/152468021/e3ce847a-31d7-4905-be10-5b0ffd40fb0f)

The deformation is defined by assigning the velocities of the points of the body by calculating the difference of x and y positions of NACA0012-NACA0006 and dividing by the time of deformation. A cubic step smoothing function is used to ease in and out the deformation. 

 
**Cd Plot: values calculated during the loop vs post the loop**
<img width="602" alt="nnnnnnn" src="https://github.com/JuliaIBPM/ViscousFlow.jl/assets/152468021/3e044ccf-ef0c-4a5b-bbc9-7b46ec2fda82">

Note that this issue arises when doing deformations. Cases with only rigid body motion didn't show this problem. Also I noticed that not all deformation cases do this.

I also monitored the change of previously calculated Cd value. In the figure below, Cd values at t = 3, 6, 9, and 12 have been plotted against time. Meaning that when the Cd is calculated at the corresponding time represented by the dot, It continues to change although the time has advanced. You can see that this occurs at all Cd values before the end of deformation at t = 10 and not after that. I have check all the values but here are 4 points to represent the issue.

![kokokokokook](https://github.com/JuliaIBPM/ViscousFlow.jl/assets/152468021/6e9b06b2-a382-49d2-a27c-9cb6e50320a3)
 
For example, I was playing with the number of airfoil points and got this

![trust1](https://github.com/JuliaIBPM/ViscousFlow.jl/assets/152468021/75f2d064-c9f5-4432-bf20-193237980623)

You can see the crazy behavior before t=10 but they all approach the expected value at the end of the deformation. To check that this is only when using the deformation, I ran the stationary case for NACA0006 with different airfoil points as shown below

![trust2](https://github.com/JuliaIBPM/ViscousFlow.jl/assets/152468021/11590d9c-a1e1-4c83-9a05-e1e293475da3)

Also, you will notice in the time loop a use of a break statement. This is because the solver seems to get stuck or crash at the last time step. So I assigned more than the intended final time and broke out at the desired time. I think this only shows in cases involving deformations but I'm not sure. 

I noticed also that the post solution has an extra time step with usually a jump in values. Thats why when plotting the "Cd_during" I used sol.t[end-1] to match the correct length. 

Sorry for the lengthy post, I hope these observations help in the debugging. Thank you again for your efforts.

**The Code**

```julia
using ViscousFlow, Plots, CSV, DataFrames

# macro to animate the motions
macro animate_motion(b,m,dt,tmax,xlim,ylim,body0,body2)
    return esc(quote
            bc = deepcopy($b)
            t0, x0 = 0.0, init_motion_state(bc,$m)
            dxdt = zero(x0)
            x = copy(x0)

            @gif for t in t0:$dt:$tmax
                motion_rhs!(dxdt,x,($m,bc),t)
                global x += dxdt*$dt
                update_body!(bc,x,$m)
                plot(bc, label="morphing", xlims=$xlim, ylims=$ylim, xlabel="x", ylabel="y", title="t = $(round(t,digits=2))", titlelocation=:left, fill=false, lc=:Red)
                plot!(body0, label="initial", fill=false, lc=:Blue, linestyle=:dash)
                plot!(body2, label="final", legend=:bottomleft, fill=false, lc=:Green, linestyle=:dash)
            end every 10
        end)
end

####################################################################################
# defining a deformation of constant velocity as a function of time
struct LinearMorphing{UT, VT} <: AbstractDeformationMotion
    u :: UT
    v :: VT
end

function RigidBodyTools.deformation_velocity(body::Body,def::LinearMorphing,t::Real)
    return vcat(def.u(t), def.v(t))
end
####################################################################################

# General Parameters
my_params = Dict()
my_params["Re"] = 200
my_params["grid Re"] = 4.0
my_params["freestream speed"] = 1.0
my_params["freestream angle"] = 0.0
my_params["CFL"] = 0.5
xlim = (-2.0, 2.0)
ylim = (-2.0, 2.0)

# Grid
g = setup_grid(xlim,ylim,my_params)
ds = surface_point_spacing(g,my_params)

# Geometry
body1 = NACA4(0.0,0.4,0.06,71) 
body2 = NACA4(0.0,0.4,0.12,71)

body0 = deepcopy(body1) # initial body

t1 = 5.0        # defomration start time
t2 = 10.0       # deformation end time
dts = 0.5       # dt for step ease in and out

Δx = body2.x̃end - body1.x̃end
Δy = body2.ỹend - body1.ỹend
Δt = (t2-t1)-dts


# Motions
dofs = [ConstantVelocityDOF(0.0)]
Xp_to_jp = MotionTransform(0.0, 0.0, 0.0)
Xc_to_jc = MotionTransform(0.0, 0.0, 0.0)
joint = Joint(RevoluteJoint, 0, Xp_to_jp, 1, Xc_to_jc, dofs)


# Deformations
# a cubic smoothing function
function CubicSmoothStep(t1, t2, t) 
    tm = maximum([0.0, minimum([1.0, (t-t1)/(t2-t1)])])
    y = tm^2*(3.0-2.0*tm)
    return y
end

# constant surface velocities as function of time
uf(t) = (Δx/Δt).*(CubicSmoothStep(t1,t1+dts,t).-CubicSmoothStep(t2-dts,t2,t)) 
vf(t) = (Δy/Δt).*(CubicSmoothStep(t1,t1+dts,t).-CubicSmoothStep(t2-dts,t2,t)) 
        
deform = LinearMorphing(uf, vf)
m = RigidBodyMotion(joint, body1, deform)

@animate_motion body0 m 0.01 15 (-0.6,0.7) (-0.5, 0.5) body0 body2

x = init_motion_state(body1, m)
update_body!(body1, x, m)

# Boundary Condition
function my_vsplus(t,x,base_cache,phys_params,motions)
    vsplus = zeros_surface(base_cache)
    surface_velocity!(vsplus,x,base_cache,motions,t)
    return vsplus
end
  
bcdict = Dict("exterior" => my_vsplus)

# System Structure
sys = viscousflow_system(g, body1, phys_params=my_params, motions=m, bc=bcdict);

# Initilizing
u0 = init_sol(sys)

tstart = 0.0
tfinal = 16.0
tspan = (tstart, tfinal)
integrator = init(u0, tspan, sys)
dt = timestep(u0, sys)

t0, x0 = 0.0, init_motion_state(body1,m)
dxdt = zero(x0)
x = copy(x0)
Cd_during = [] # an empty array to hold Cd values during the loop

# Solving
@time "Global" for t in tstart:dt:tfinal
    if t>15
        break
    end
    print("t = ")
    print(t)
    motion_rhs!(dxdt,x,(m,body1),t)
    global x += dxdt*dt
    update_body!(body1,x,m)
    @time step!(integrator, dt)
    local mom, fx, fy = force(integrator.sol,sys,t,1);
    push!(Cd_during, 2*fx)
end

# Post Processing
sol = integrator.sol;
mom, fx, fy = force(sol,sys,1);
Cd_post = 2*fx;

# steady state stationary Cd values for NACA0006 and NACA00012 for Re = 200, Grid Re = 4.0 and Domain 4x4
Cd_NACA0006 = 0.27823
Cd_NACA0012 = 0.29718

# plotting
plot(sol.t,Cd_post,label="post",xlim=(0,Inf),ylim=(0.1,0.5),xlabel="Convective time",ylabel="\$C_D\$", lc=:Red)
plot!(sol.t[1:end-1],Cd_during,label="during",xlim=(0,Inf),ylim=(0.1,0.5),xlabel="Convective time",ylabel="\$C_D\$", lc=:Blue)
hline!([Cd_NACA0006], label="NACA0006", linestyle=:dash, lc=:Black)
hline!([Cd_NACA0012], label="NACA0012", linestyle=:dash, lc=:Black)

```

**Comment by jdeldre** [comment]
Ah, I found the error. I hadn't seen it before because I haven't done many tests with deforming bodies. Fixing it now...

**Comment by jdeldre** [comment]
I think that this issue is now resolved by PR #94. Please try it out in the newly-released version.

Also, just a suggestion: since it takes longer to run when a deforming motion is applied, but you don't start deforming until late in the simulation, you might consider running with a static motion until just before the deformation starts. This will be much faster. Then, you can use the final state of this as your initial condition for the deforming motion. The only caveat is that you have to create a new auxiliary state (x) for the initial condition of the deforming problem.

```
ui = deepcopy(integrator.u)
ti = integrator.t
xi = init_motion_state(body1, m)
u0 = solvector(state=state(ui),constraint=constraint(ui),aux_state=xi)
```

**Comment by mxalhussaini** [comment]
Thank you for the fast response and quick fix. Yes the new version fixed the problem as shown below.

<img width="599" alt="Screen Shot 2023-12-01 at 6 16 40 PM" src="https://github.com/JuliaIBPM/ViscousFlow.jl/assets/152468021/0d6b7e64-9c68-4b0e-b0cf-7efe7ccc60b8">


Also, great suggestion; Thank you for clarifying what needs to be done to implement it.

I think it would be helpful if one example of a deformation case is added to the examples in the documentation of ViscousFlow. I only found how to define deformations and how to animate them in the RigidBodyTools documentation but not how to solve using them in ViscousFlow. 

Thank you again, great work!
